thunder/core/jit_ext.py

```diff
@@ -617,6 +618,231 @@ def _general_jit_torch_finfo_lookaside(dtype: thunder.dtypes.dtype):
     return res
 
 
+from torch.utils.checkpoint import noop_context_fn
+
+
+@register_general_jit_lookaside(torch.utils.checkpoint.checkpoint)
+@register_general_jit_lookaside(torch.ops.higher_order.tag_activation_checkpoint)
+def _general_jit_torch_checkpoint_lookaside(
+    function: Callable,
+    *args,
+    # use_reentrant=None,
+    # context_fn=noop_context_fn,
+    # determinism_check="default",
+    # debug=False,
+    **kwargs: Any,
+):
+    """
+    This function does preprocessing of the `function` argument before
+    dispatching the call to `thunder.torch.checkpoint`. This is necessary
+    because the `function` is potentially calling into PyTorch functions that
+    are not yet translated to Thunder. `thunder.torch.checkpoint` is a Thunder
+    function that can handle only Thunder functions as input.
+
+    Args:
+        function: The function to be checkpointed.
+        args: Arguments to the function.
+        kwargs: Keyword arguments to the function.
+
+    Returns:
+        The result of calling `thunder.torch.checkpoint` with the preprocessed
+        `function` and its arguments.
+    """
+    # from thunder.torch import checkpoint
+    from thunder.core.baseutils import check, sequencify
+    from thunder.core.transforms import augmented_forward_impls, backward_impls, VJPDual
+
+    jit_ctx: JitCtx = get_jit_ctx()
+
+    # Construct computation trace(trace_of_checkpoint), checkpoint_fwd_sym
+    jit_ctx.computation_trace.push_scope([])
+    func = unwrap(function)
+    result = _interpret_call(func, *args, **kwargs)
+
+    if result is INTERPRETER_SIGNALS.EXCEPTION_RAISED:
+        return result
+
+    bsyms: list[BoundSymbol] = jit_ctx.computation_trace.pop_scope()
+    unwrapped_result = unwrap(result)
+    trace_of_checkpoint = TraceCtx()
+    for bsym in bsyms:
+        trace_of_checkpoint.add_bound_symbol(bsym)
+    with tracectx(trace_of_checkpoint):
+        prims.python_return(unwrapped_result)
+
+    unwrapped_args = tree_map(lambda a: unwrap(a), args)
+
+    si = SigInfo("activation_checkpoint")
+    si.args.append(("function", None))
+    for a in unwrapped_args:
+        if isinstance(a, Proxy):
+            si.args.append((a.name, None))
+        else:
+            pa = proxy(a)
+            si.args.append((pa.name, None))
+    trace_of_checkpoint._siginfo = si
+    trace_of_checkpoint.args = (func, *unwrapped_args)
+
+    @wraps(trace_of_checkpoint.python_callable())
+    def core_of_forward(f, *args, **kwargs):
+        return thunder.core.trace_interpreter.interpret_trace(trace_of_checkpoint, f, *args, **kwargs)
```

Comment:
Can this `core_of_forward` (but without `f` argument) be sent as the function-to-be-checkpointed to `thunder.torch.checkpoint` function from https://github.com/Lightning-AI/lightning-thunder/pull/1127? Then the rest of the custom registration could probably be dropped.